add_codeowners.py

```diff
@@ -0,0 +1,498 @@
+
+
+@dataclass
+class RepoOwnerSettings:
+    owner: str
+    repo_name: str
+    sme_owners: List[CodeownerGroup]
+    lang_owners: List[CodeownerGroup]
+    area_owners: List[CodeownerGroup]
+
+    @classmethod
+    def from_dict(cls, source: dict):
+        settings = cls(
+            owner=source["owner"],
+            repo_name=source["repo_name"],
+            sme_owners=[CodeownerGroup(value.strip()) for value in
+                        source["sme_owners"].split(',') if value],
+            area_owners=[CodeownerGroup(value.strip()) for value in
+                         source["area_owners"].split(',') if value],
+            lang_owners=[CodeownerGroup(value.strip()) for value in
+                         source["lang_owners"].split(',') if value],
+        )
+
+        settings.sme_owners.sort()
+        settings.area_owners.sort()
+        settings.lang_owners.sort()
+
+        return settings
+
+
+GROUP_TO_EXT: Dict[CodeownerGroup, str] = {
+    CodeownerGroup.PYTHON: '*.py*',
+    CodeownerGroup.TWINCAT: '*.{plcproj,sln,TcDUT,TcGVL,TcPOU,TcTTO,tmc,tsproj,xti}',
+    CodeownerGroup.CPP: '*.cpp',
+    CodeownerGroup.SHELL: '*.sh',
+    CodeownerGroup.EPICS: '*.{archive,autosave,cmd,db,dbd,edl,ioc,proto,req,'
+                          'sub-arch,sub-req,substitutions,tpl-arch,tpl-req}',
+}
+
+
+def create_codeowners_file(settings: RepoOwnerSettings) -> str:
+    """
+    Create codeowners file based on RepoOwnerSettings
+
+    See https://confluence.slac.stanford.edu/x/u5Q9Hw for details.
+    By default:
+    - "default" = sme_owners + area_owners
+    - .github owned by admins only
+    - Default (*): default
+    - language specific files: default + lang_owners
+
+    Returns content encoded as base64 utf-8, in order to ensure the
+    payload is appropriate for github's REST API
+    """
+    lines = []
+
+    # gather groups
+    default_owners = list(set(settings.sme_owners + settings.area_owners))
+    if not default_owners:
+        default_owners = [CodeownerGroup.ADMIN]
+    default_groups = ["@pcdshub/" + grp for grp in default_owners]
+
+    # .github admin group
+    lines.append("# github folder holds administrative files")
+    lines.append(f".github/** @pcdshub/{CodeownerGroup.ADMIN}")
+    lines.append("")
+
+    # default group
+    lines.append("# default group")
+    lines.append(f"* {' '.join(default_groups)}")
+    lines.append("")
+
+    # Language specific groups
+    if settings.lang_owners:
+        lines.append("# language-specific group(s)")
+        for group in settings.lang_owners:
+            specific_lang_group = ' '.join([f"@pcdshub/{group}"] + default_groups)
+            lines.append(f"{GROUP_TO_EXT[group]} {specific_lang_group}")
+        lines.append("")
+
+    base_file = "\n".join(lines)
+
+    # ensure file contents can be base64 encoded for submission to github API
+    encoded_content = base64.b64encode(base_file.encode('utf-8'))
+    return encoded_content.decode("ascii")
```

Comment:
Question:
What's up with encoding as utf-8 and decoding as ascii? Does this have something to do with the specifics of base 64 encoding?

Reply:
Good catch 👍 

If I'm honest this was the magic that worked and I stopped looking at it out of frustration.  (I'm less frustrated now 😆 ) The key here is that `base64.b64encode` returns a byte-string, and we need to pass `gpapi` a normal string.  decoding as `utf-8` should work too, and I've just verified that it does.